Switch to using the Core local environment

pento opened the pull request:
## Description

Now that WordPress Core has its own Docker environment, it's possible to base the Gutenberg environment on that.

There are several advantages for Gutenberg, when compared to the existing environment:

- It's cross-platform, only requiring Docker to be installed and running. It also works with Docker Toolbox, allowing all Windows users (not just those with Windows 10 Pro) to use it.
- The WordPress environment uses Docker images that we maintain for the purpose of Core development, which allows us to to customise them as we need.
- It ensures that tests for Core and Gutenberg run in the same environment.
- It's a single environment, which helps avoid the naming clashes we've occasionally seen between the Core and Gutenberg environments.

More broadly for WordPress, tying the development processes of Core and Gutenberg a little closer together helps pave the way for the processes to merge over time.

## Architectural Aims

This PR removes the Docker config from Gutenberg entirely (barring a minimal template to hook into the Core environment): requiring `wordpress-develop` to run is intentional.

There are a few reasons for requiring `wordpress-develop`. It's worth noting that one of my primary guides for making decisions around `local-env` is to lay the groundwork for a dev environment that works as easily as possible for non-devs. With that in mind:

* Having a single `docker-compose.yml` file in Core, as opposed to having one in Core, one in Gutenberg, and many more in other feature plugins that will be able to use this, means that there's only set of Docker containers running. Docker can get fairly weird if you're trying to run containers with the same names, which mount volumes with the same names.
* Having multiple `docker-compose.yml` files doesn't lend itself to testing plugins together: each plugin ends up with its own environment which can't talk to others.
* Managing how plugins can mount older versions of WordPress becomes exponentially more difficult if the `docker-compose.yml` config needs to work for every WordPress branch (this is a requirement for `local-env` so that auto updates can be worked on in old branches), rather than being able to tweak the config for each branch.
* Ideally, feature plugins shouldn't need to do much customisation of the Docker environment, beyond mounting their particular volumes. Exposing the entire `docker-compose.yml` file is overkill.

That said, not every Gutenberg contributor will have a WordPress source clone, or want to manage one. So, there's also a helper which will download and install a copy of the WordPress source repo, when needed. 

Much like the WordPress environment, however, I've aimed to _not_ automate everything. As we found with the Gutenberg environment, too much automation tended to make it difficult (particularly for more advanced contributors) to customise or debug their environment. For example, the scripts no longer install NVM, switch your Node version, or run `npm install`.

## How It Works

If you don't have a WordPress repo to work with, just run `npm run env install`, that'll take care of downloading WordPress, building it, and connecting Gutenberg into it.

Alternatively, setting the `WP_DEVELOP_DIR` environment variable to your WordPress source directory, and running `npm run env connect` will write a `docker-compose.override.yml` file to the WordPress source directory, and restart the Docker containers. The Docker override file contains the appropriate volume settings for mounting Gutenberg within the WordPress environment.

Here's a complete list of the commands added in this PR:

`npm run env install`: Automatically downloads, builds, and installs a copy of WordPress to work with. This will be installed in the `wordpress` folder.

`npm run env connect`: If you have a WordPress respository checked out already, define the `WP_DEVELOP_DIR` environment variable with the path to your repository, then run this command to mount Gutenberg inside the Core environment.

`npm run env start` / `npm run env stop`: After the environment is setup, there's no need to run the install script again. the `start` and `stop` scripts will start and stop the Docker containers.

`npm run env update`: When a managed WP install is being used, the `update` script will update the WordPress install and rebuild it. This doesn't happen automatically, but the `start` script will show warnings if it hasn't been run for more than 7 days.

`npm run env reinstall`: Resets the database and re-configures WordPress again.

`npm run env cli`: This is a shortcut for WP-CLI, it also ensures that the CLI command is being run against `src` or `build`, depending on what you've set in `LOCAL_DIR`.

`npm run env docker-run`: This is a shortcut for `docker-compose run --rm`.

`npm run env test-php`: This will run the PHPUnit tests. 

`npm run env lint-php`: This will run the composer lint scripts.

## Testing

After both of these methods, your WordPress install will be at http://localhost:8889/wp-admin/, username is _admin_, and password is _password_.

### Method A (download and configure WordPress automagically)

For testing the managed WordPress install, just run `npm run env install`, and watch the magic happen. 🎉

### Method B (connect to an existing `wordpress-develop` install)

If you want to test with an external WordPress directory, check it out, and run these commands in your WordPress directory:

```sh
nvm install
npm install
npm run build:dev
npm run env:start
npm run env:install
```

Then in your Gutenberg directory, run:

```sh
export WP_DEVELOP_DIR=/path/to/wordpress-develop
npm run env connect
npm run env cli plugin activate gutenberg
```

## Reusability

These scripts are intended to be reusable by other Core feature plugins. While they may support third party development in the future, this isn't an immediate goal.

pento committed:
First pass at use core local env.

pento committed:
Delete all the files this setup will replace. (Many of them are not yet replaced, let's see what explodes!)

pento committed:
Retrieve the compose file version from Core's docker-compose.yml, to ensure the override version matches.

pento committed:
Change NPM scripts to use the Core Docker environment.

pento committed:
Use local-env in Travis.

pento committed:
Run composer locally, instead of in Docker.

pento committed:
Run `npm install` for WordPress in Travis.

pento committed:
Upgrade Docker on Travis

pento committed:
Use the right build script name.

pento committed:
Travis needs to `npm ci` for all jobs now.

pento committed:
push !== append

pento committed:
Run `composer install` in the PHP Unit Test jobs.

pento committed:
Add debugging

pento committed:
Change the COMPOSER env var to COMPOSER_INSTALL. The former is used by composer itself.

pento committed:
Build Gutenberg before running it.

pento committed:
Fix the PHPUnit multisite test command.

pento committed:
Activate the Gutenberg plugin after installing it.

pento committed:
Run e2e tests against the build dir.

pento committed:
Build to the build dir

youknowriad commented on .travis.yml:
```diff
@@ -24,14 +24,49 @@ branches:
+      sudo mv docker-compose /usr/local/bin
+
+      # Check out WordPress
+      git clone git://develop.git.wordpress.org/ wordpress
```
I don't like the fact that we're relying on the local environment of an external repository to achieve that. Especially since it's not written to be reused outside of Core as a reusable "module/package/script". 

Can't we instead, just have the docker-compose file here in this repo too (copied).
Also, I kind of like  the fact that when you open `package.json` you see the scripts `env:start`.... I feel we should have these scripts in this repo tooo.

youknowriad reviewed (dismissed):
The `./bin/setup-local-env.sh` script is documented in a lot of places of the repo. Contributors  shouldn't have to clone another repository to trigger a local environment to work on Gutenberg.

We should instead have a local environment embedded in this repository instead. Sharing the same config as Core makes sense.

pento commented:
Thank you for your initial thoughts, @youknowriad. This PR is still highly experimental, I haven't touched documentation or worked on polishing the UX, since how it all works together is likely to change as I continue to explore it.

Similarly, you're likely to come across fairly hacky bits and pieces which aren't really ready for review. (For example, Travis config is currently hacked together for looking at "what would it take to get Travis passing in this scenario".)

youknowriad commented:
Oh sorry @pento for being too fast. Let me know when you need feedback

pento committed:
Gutenberg is in the build directory on Travis.

pento committed:
Travis debugging.

pento committed:
Run PHP unit tests in the build dir.

pento committed:
Configure WordPress to load from the build directory.

pento committed:
Run author e2e tests as an author.

pento committed:
Fix the command for setting up the author user.

pento committed:
Travis: create the uploads directory.

pento committed:
Travis: Don't need to disable SCRIPT_DEBUG for e2e tests.

pento committed:
Move the env scripts to wp-scripts.

pento committed:
Add a bunch of env helper scripts.

pento committed:
Travis: Composer tasks need WP installed, now.

pento committed:
Fix the mount directory in the PHPUnit container.

pento committed:
Don't lint CSS in the WordPress directory.

pento committed:
Disable SCRIPT_DEBUG, and clean up the globals.

pento committed:
Generate docker-compose.override.yml from a template file.

peterwilsoncc commented on .gitignore:
```diff
@@ -10,7 +10,6 @@ gutenberg.zip
 *.log
 phpcs.xml
 yarn.lock
-docker-compose.override.yml
```
Probably worth keeping this around:

* it's benign
* once a user pulls these commits, their override file will still be in place so an accidental commit is bordering on likely.

pento replied:
I'm inclined to remove it, as while it should be benign, it'd be the source of the weirdest bugs in existence if it turns out that it does clash in some way. It's better to ensure the file is deleted, rather than let it hang around.

Even if an accidental commit happens, I don't think a PR with the accidental commit in it would be merged.

pento commented:
> This does have the potential drawback of requiring Gutenberg contributors who may be unfamiliar with WordPress Core development to check out and build a copy of WordPress. I haven't written a script to automate this task, but I'm open to arguments for or against it. See the before_install section of .travis.yml for an example of the steps needed in an automated solution.

After some discussion over in the core-editor channel, there's clearly still a need for this to be automated in some form.

pierlon commented:
Instead of cloning `wordpress-develop`, could it be referenced as a submodule?

pento commented:
@pierlon: The Docker config only lives in the `wordpress-develop` repo, it isn't included in the WordPress zip download.

A shallow clone is very fast, but the downside is that it requires `git` to be installed, and available in the path. I'll look at grabbing the `wordpress-develop` zip from GitHub, as an alternative, but that has its own issues.

pento committed:
Add start/stop scripts.

pento committed:
Add an install script.

pento committed:
Tidy up .travis.yml

pento committed:
Travis: WordPress install now needs to happen after npm ci.

shivapoudel commented:
@pento Is it possible to base the custom Xyz plugin environment on core like Gutenberg? If yes how can I modify wp-scripts docker override YAML file as it has hardcoded Gutenberg word?

pento committed:
Allow other plugins using these scripts to override the docker-compose template with their own.

pento commented:
Thank you for reminding me, @shivapoudel! I half-implemented this, but didn't get back to it. 🙂 5e1c8c2 adds the ability to override the default `docker-compose.override.yml.template` with your own, by adding this section to your `package.json` file:

https://github.com/WordPress/gutenberg/pull/17004/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R237-R240

pento committed:
Try activating the plugin using the core WP-CLI script.

pento committed:
Run the correct core build command.

shivapoudel commented:
@pento I have some notes for this PR enhancement from my side as I am working on a custom plugin and implementing this scripts :)

If the second try of `npm run env install` is executed. I receive this error 
![image](https://user-images.githubusercontent.com/3774827/63246883-673bfe00-c283-11e9-80e3-b7fa06828cc5.png)
Maybe check if a folder `wordpress` exists and build is done OR just go inside and pull the repo. Remember I don't want to set `WP_DEVELOP_DIR` in custom plugin so folder is inside plugin.

Additionally, is there any possibilities to change the site title, email used from this script. As a custom plugin, I want to set different site title and email address. Finally, how can I set the build version of `wordpress` rather than src being in current plugin folder?

pento committed:
b0d7330cf773d06a292a56ab9cb27793a7874c39 is obseleted by #16243.

shivapoudel commented:
@pento I am in a windows environment with Git bash and MySQL volume is not synced in my WordPress directory (which is inside the plugin) due to the format of [this line](https://github.com/WordPress/wordpress-develop/blob/23769b62c12d3d7b603f391f885cb91c0eb72aa3/docker-compose.yml#L64) :)

If I personally change this format into `- ./mysql:/var/lib/mysql` then it syncs the MySQL volume perfectly. Maybe the culprit for windows user but honestly if you have much more idea about where it syncs then it will be appreciated.

pento committed:
Check if WordPress is already installed before trying to install it again.

pento commented:
@shivapoudel: Thank you for the testing and feedback!

0b4c7f5 adds a check to see if WordPress is already installed.

Good point on old volumes staying around when you run `npm run env connect` after removing them from the template. I'll see what can be done about that.

> Is there any possibilities to change the site title, email used from this script. As a custom plugin, I want to set different site title and email address.

I think we can explore this, but it's a lower priority at the moment. Right now, the aim is to build something that works for Core and Feature Plugin development. If it works for your own plugin development, that's awesome, but I'd like to keep the focus on adding options that Core/Gutenberg need, at least for now. 🙂

> Finally, how can I set the build rather than src version of wordpress while being in the plugin directory, not in the WP develop core directory and setting up it .env file?

It will recognise the `WP_DEVELOP_DIR` environment variable for this. Run `export WP_DEVELOP_DIR=build` before you run `npm run env connect` to set it all up. this will only effect your current terminal window, though, we can explore making it possible to set this for the entire project.

Finally, will there be a WP version selection OR development version only as a future concise asking it here? If so will the previous version of WP includes docker-compose.yml or not :)

There will be in the future, but WP branches currently don't have the Docker config. I'm planning on backporting it once it's a bit more stable. 🙂

> If I personally change this format into - ./mysql:/var/lib/mysql then it syncs the MySQL volume perfectly. Maybe the culprit for windows user but honestly if you have much more idea about where it syncs then it will be appreciated.

It's referring to the `mysql` [further down](https://github.com/WordPress/wordpress-develop/blob/23769b62c12d3d7b603f391f885cb91c0eb72aa3/docker-compose.yml#L115) in the `docker-compose.yml` file. I put the MySQL files in a Docker volume so that they persist between restarts, but can be easily removed (for example, if you're switching to an older version of MySQL, it's easier to work with a fresh install, rather than downgrading the data file formats).

shivapoudel commented:
@pento Thanks you for your reply with such insight... What about the second time run of `npm run env connect` as the linked docker override template might change. After the install of the environment if I again try to run connect then docker override is not updated correctly!

pento commented:
@shivapoudel: Could you clarify what's going on here? Are you able to reproduce the same problem by modifying the Gutenberg `bin/docker-compose.override.yml.template` file, and specifying how you would expect it to behave?

pento committed:
If we're running in a managed environment, encourage updating if it hasn't been updated in over a week.

shivapoudel commented:
@pento I have created a gist to explain what is happening in a better way. Please follow this step:

After executing `npm run env install` **docker-compose.override.yml** file is generated alright! That file looks perfect and no more confusion. But during the development process, I might need other volumes to be synced so I modify the docker template file linked via `package.json` by including all volumes and try to execute the command `npm run env connect` which now updates docker override file with wrong volumes information. **BUT** if I delete the override file located inside `wordpress` directory and try to execute the connect command again then correct information is there in the override file after the update. Please see [this gist](https://gist.github.com/shivapoudel/3cc7fa660af711d54e451fef69c4f641) files for what config was set in the docker override file if that already exists and doesn't exist inside `wordpress` directory :)

Please remember this script was executed in the custom plugin but not in Gutenberg.

pento committed:
Check the Node version before trying to do any environment work.

pento committed:
Add a placeholder for setup-local-env.sh, pointing folks at the new way of doing things.

pento committed:
Update the documentation for the local-env.

pento committed:
Add inline commenting for mergeConfigs().

pento committed:
Ensure we catch invalid YAML template files.

pento committed:
Improve behaviour on Windows.

pento commented:
Thanks, @shivapoudel! I was able to reproduce the issue, it's now fixed as of c447844cb.

pento commented:
This PR is getting pretty close to merge-able. I'd appreciate some reviews! 🙂

shivapoudel commented:
@pento Thanks for the fix its butter-smooth :)

 One more question, if previously mounted volumes are renamed from docker template then both previous and renamed one exists in the volumes list. Somehow seeing the process looks like this is intentional but can we expect something like `npm run env connect --force` or something like that command to make a fresh copy rather than merges :)

noisysocks commented:
Some loose thoughts from a very quick and not at all thorough test:

- It would be nice if running `npm run env` with no arguments listed the available commands with a short description.
- It would be helpful if `npm run env start` displayed a friendly _You might want to run `npm run env install.`_ message after its error about `WP_DEVELOP_DIR` not being set.
- Lots of folks will be in the habit of running `./bin/setup-local-env.sh`. I suggest we deprecate that script rather than removing it. It could print out instructions on what the new setup is (`npm run env install` and `npm run env start`), or just run those two commands with a warning.
  - Similarly, a lot of folks will be in the habit of going to `http://localhost:8888` instead of `http://localhost:8889` 😛
- I noticed we're closing the `master` branch of `wordpress-develop`. I'm wondering if this breaks anything. In particular, do our PHP shims in `lib/compat.php` work when the plugin is installed on WordPress trunk? (I've never checked!)
- I can't get `npm run test-php` to work. It fails with `ERROR: The /build/ directory is missing! Please run `npm run build` prior to running PHPUnit.`
- Love that there's an easy way to run WP CLI! This will be really useful for day-to-day testing and development 🤤

Excited about all the 🔻 here! I'll take a proper look at everything soon 🙂

pento commented:
> One more question, if previously mounted volumes are renamed from docker template then both previous and renamed one exists in the volumes list.

That's a good point, thanks @shivapoudel. I think the current method tries to be a little too clever when merging, it can be simplified a bit.

Thank you for the suggestions, @noisysocks! I like all of them, I'll get to work on implementing them. 🙂

pento committed:
Simplify the array merging, so old items aren't left hanging around.

pento committed:
Add an extra prompt to help people who run env commands without a WP_DEVELOP_DIR set.

pento committed:
List the commands when no arguments are passed to npm run env.

pento committed:
Auto switch between running PHPUnit tests from the src or build directory.

shivapoudel commented:
@pento Thanks! That works great :)

pento committed:
Fix PHPUnit tests not properly running in the build directory.

pento committed:
Let's keep the welcome message when installing local-env.

youknowriad reviewed (commented):
I checked the docs update and it gives it clear idea of how this PR updates the environment. I think it's great. Thanks @pento for the updates.

I can't stop but think that, all these scripts and setup would be good to "share" somehow (not necessarily a package) for any plugin to use.

pento commented:
> I can't stop but think that, all these scripts and setup would be good to "share" somehow (not necessarily a package) for any plugin to use.

I agree, sharing them could be a potential future step to help standardise plugin development. As @shivapoudel has been experimenting with, this is already possible, though doesn't yet provide all the customisability that the entire plugin ecosystem would need.

For now, the focus is on something that works for Core, Gutenberg, and other feature plugins.

noisysocks commented on packages/scripts/scripts/env/install.js:
```diff
@@ -0,0 +1,97 @@
+/**
+ * External dependencies
+ */
+const commandExistsSync = require( 'command-exists' ).sync;
```
We can make this look a little closer to `import` syntax.

```suggestion
const { sync: commandExistsSync } = require( 'command-exists' );
```

noisysocks commented on packages/scripts/scripts/env/start.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * External dependencies
+ */
+const commandExistsSync = require( 'command-exists' ).sync;
```
We can make this look a little closer to `import` syntax.

```suggestion
const { sync: commandExistsSync } = require( 'command-exists' );
```

noisysocks commented on packages/scripts/scripts/env/stop.js:
```diff
@@ -0,0 +1,10 @@
+
```
Extra `\n` here.

```suggestion
```

noisysocks commented on packages/scripts/scripts/env/update.js:
```diff
@@ -0,0 +1,61 @@
+/**
+ * External dependencies
+ */
+const commandExistsSync = require( 'command-exists' ).sync;
```
We can make this look a little closer to `import` syntax.

```suggestion
const { sync: commandExistsSync } = require( 'command-exists' );
```

noisysocks commented on packages/scripts/scripts/env/update.js:
```diff
@@ -0,0 +1,61 @@
+	stdout.write( 'Downloading...\n' );
+	// Download the archive.
+	request( 'https://github.com/WordPress/wordpress-develop/archive/master.zip' ).pipe( tmpZipWriter );
+}
```
This is the same as what's in `scripts/env/install.js`. Would it make sense to refactor it out into a `util` function?

pento replied:
Refactoring is fun for the whole family! 833f76328

noisysocks commented on packages/scripts/scripts/env/install.js:
```diff
@@ -0,0 +1,97 @@
+	tmpZipWriter.on( 'finish', () => {
+		const unzipper = new DecompressZip( tmpZip );
+
+		unzipper.on( 'extract', buildWordPress );
```
Is there an `error` event here that we should be handling?

pento replied:
Added as part of 833f76328.

noisysocks commented on packages/scripts/scripts/env/install.js:
```diff
@@ -0,0 +1,97 @@
+
+	stdout.write( 'Downloading...\n' );
+	// Download the archive.
+	request( 'https://github.com/WordPress/wordpress-develop/archive/master.zip' ).pipe( tmpZipWriter );
```
Is there an `error` event here that we should be handling?

pento replied:
Added as part of 833f76328.

noisysocks commented on .travis.yml:
```diff
@@ -22,20 +21,55 @@ branches:
+    - LOCAL_DIR: build
+    - LOCAL_SCRIPT_DEBUG: false
+    - COMPOSER_INSTALL: false
+    - WORDPRESS: true
```
It wasn't immediately clear to me what `WORDPRESS` does, could we name it `INSTALL_WORDPRESS` or something? (Could rename `COMPOSER_INSTALL` to `INSTALL_COMPOSER`, too, so that there's parity.)

pento replied:
👍🏻 5c43925a5

noisysocks commented on .travis.yml:
```diff
@@ -62,115 +94,74 @@ jobs:
-      language: php
-      php: 5.6
-      env: WP_VERSION=latest
+      env: COMPOSER_INSTALL=true LOCAL_PHP=5.6-fpm
```
This is so much nicer! 😍

pento replied:
💖

noisysocks commented on .travis.yml:
```diff
@@ -62,115 +94,74 @@ jobs:
-    - name: PHP unit tests (PHP 5.2)
-      env: WP_VERSION=latest SWITCH_TO_PHP=5.2
-      script:
-        - ./bin/run-wp-unit-tests.sh
```
Good riddance! 🎉

pento replied:
😃

noisysocks commented on .travis.yml:
```diff
@@ -22,20 +21,55 @@ branches:
     - master
 
+env:
+  global:
```
I'd prefer that the jobs always specify `COMPOSER_INSTALL` and `WORDPRESS` so that the reader doesn't have to scroll up and think about what the default is.

noisysocks replied:
Eh, actually, it's fine. 🙂

pento replied:
I found it was getting pretty messy and repetitive if every job specified the same default values.

noisysocks reviewed (approved):
Great job on this! I love that this PR decimates what's in `bin` and simplifies `.travis.yml`.

Everything worked as expected in my testing, bearing in mind that I did **not** test on Windows. None of my comments are blocking, so giving this the ✅.

pento commented on package.json:
```diff
@@ -236,5 +237,18 @@
+			"`---' `---' `---’ `---’ '   ` `---' `---’ `     `---|",
+			"                                                `---'"
+		],
+		"welcome-build-command": "npm run dev"
```
So they can be customised by other plugins that use these scripts. I've tried (I think succeeded) to remove all Gutenberg-specific references or code from the scripts in `wp-scripts`.

noisysocks replied:
Nice, that makes sense 👍

pento commented on packages/scripts/package.json:
```diff
@@ -41,16 +41,21 @@
+		"request": "2.88.0",
 		"resolve-bin": "^0.4.0",
 		"source-map-loader": "^0.2.4",
+		"sprintf-js": "^1.1.1",
```
I don't know why all of the `package.json` files in `packages` use mixed notations. I went with what appears to be the most common.

ntwb replied:
For the `@wordpress/scripts` package using `^` caret makes sense, it looks like all `dependencies` are using the `^` for _minor_ semver updates to take place 👍🏼

gziolo replied:
All of them should use `^`. There is only one exception `webpack` which needs to be upgraded to the latest version first which might require some config changes.

pento committed:
Normalise WP_DEVELOP_DIR when it's generated, so we don't need to normalize it everywhere it's used.

pento commented on packages/scripts/scripts/env.js:
```diff
@@ -0,0 +1,61 @@
+
+if ( ! env.WP_DEVELOP_DIR && command !== 'install' ) {
+	if ( existsSync( normalize( cwd() + '/wordpress/wp-config-sample.php' ) ) ) {
+		env.WP_DEVELOP_DIR = cwd() + '/wordpress';
```
👍🏻 f682a029c

pento committed:
Fix an incorrect comment referring to PHPUnit instead of WP-CLI.

pento commented on packages/scripts/scripts/env/cli.js:
```diff
@@ -0,0 +1,18 @@
+
+const localDir = env.LOCAL_DIR || 'src';
+
+// Run PHPUnit with the working directory set correctly.
```
👍🏻 25b8f4ae8

pento committed:
Expand abbreviated variable names.

pento commented on packages/scripts/scripts/env/connect.js:
```diff
@@ -0,0 +1,115 @@
+if ( existsSync( composeFile ) ) {
+	try {
+		compose = yaml.safeLoad( readFileSync( composeFile, 'utf8' ) );
+	} catch ( e ) {
```
👍🏻 f8d7ef3a5

pento committed:
Remove an unnecessary normalize()

pento commented on packages/scripts/scripts/env/connect.js:
```diff
@@ -0,0 +1,115 @@
+	exit( 1 );
+}
+
+const pluginMountDir = normalize( cwd() );
```
👍🏻 e37b06b9a

pento committed:
Use Array.isArray() to test for arrays.

pento commented on packages/scripts/scripts/env/connect.js:
```diff
@@ -0,0 +1,115 @@
+		if ( ! originalConfig[ key ] ) {
+			// If the originalConfig object doesn't have this element, we can just add it.
+			originalConfig[ key ] = newConfig[ key ];
+		} else if ( newConfig[ key ] instanceof Array ) {
```
👍🏻 aecf45392

pento committed:
Use isPlainObject() to test for objects.

pento commented on packages/scripts/scripts/env/connect.js:
```diff
@@ -0,0 +1,115 @@
+
+			// Append the newConfig to the remaining config.
+			originalConfig[ key ] = [ ...cleanOriginal, ...newConfig[ key ] ];
+		} else if ( newConfig[ key ] instanceof Object ) {
```
Let's play it safe and do a safer check. 3f1a380e0

pento committed:
Make require look a little more like import.

Co-Authored-By: Robert Anderson <robert@noisysocks.com>

pento committed:
Make require look a little more like import.

Co-Authored-By: Robert Anderson <robert@noisysocks.com>

pento committed:
Make require look a little more like import.

pento committed:
Remove a bonus \n

Co-Authored-By: Robert Anderson <robert@noisysocks.com>

pento committed:
Travis: Rename some environment variables, to make their meaning a little clearer.

pento committed:
Refactor repetitive code into util functions.

noisysocks commented on packages/scripts/utils/env.js:
```diff
@@ -0,0 +1,162 @@
+ * @param {string} baseDir        The base directory of the plugin, volumes in this directory belong to the plugin.
+ * @return {Object} The merged config object.
+ */
+function mergeYamlConfigs( originalConfig, newConfig, baseDir ) {
```
YAML is an acronym so this should be `mergeYAMLConfigs`. (I know, I know.)

https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#abbreviations-and-acronyms

pento replied:
🙃 4d9bde963

noisysocks reviewed (approved):
Just a naming thing and then we're 👍👍👍

pento committed:
s/mergeYamlConfigs/mergeYAMLConfigs/g

gziolo commented:
It looks like this PR still needs some performance optimizations on Travis

This branch

![Screen Shot 2019-08-26 at 12 44 02](https://user-images.githubusercontent.com/699132/63685330-53524800-c7ff-11e9-9ddf-2220bb095367.png)

The most recent run on Travis – PR #16796

![Screen Shot 2019-08-26 at 12 44 31](https://user-images.githubusercontent.com/699132/63685368-69f89f00-c7ff-11e9-943c-b78954d55f33.png)

gziolo commented on package.json:
```diff
@@ -205,20 +205,21 @@
 		"publish:dev": "npm run build:packages && lerna publish --npm-tag next",
 		"publish:prod": "npm run build:packages && lerna publish",
 		"test": "npm run lint && npm run test-unit",
-		"pretest-e2e": "cross-env SCRIPT_DEBUG=false ./bin/reset-local-e2e-tests.sh",
```
What happens to the database after all e2e tests finish their execution?

This command was used to reset all created posts, taxonomies, etc so they don't have an impact on the result of subsequent runs of the same tests.

pento replied:
Thanks for pointing that one out, @gziolo! 009f02c72 implements an equivalent reset before running e2e tests.

gziolo commented on packages/scripts/scripts/env.js:
```diff
@@ -0,0 +1,61 @@
+
+if ( ! args.length ) {
+	const scripts = {
+		install: "If you don't have a check out of the WordPress repository that you want to work with, this will automatically download, configure, and connect to WordPress.",
```
Do you plan to document this new group of scripts in the README file of `@wordpress/scripts` package as part of this PR?

pento replied:
👍🏻 7fe5c3d7a

gziolo commented:
![Screen Shot 2019-08-26 at 13 08 25](https://user-images.githubusercontent.com/699132/63687559-3751a500-c805-11e9-81aa-1a125687b54b.png)

Is this still up to date? It looks like this could be simplified for Gutenberg developers and bundled in one command similar to what we have as of today with `./bin/setup-local-env.sh`.

gziolo commented:
I'm still processing all the changes proposed. I was AFK for more than a week and I probably missed a lot of discussions related to this PR.

I should probably start with **major kudos** for all the work done here. It's amazing to see how it shapes and I wanted to emphasize that it's the last missing bit which is necessary to make plugins development smooth with all the recent introduction of modern JavaScript.

Now, I want to apologize if you have to answer the same questions over again, but here is what I still try to understand.

1. Why do you need to use `docker-compose.yml` from `WordPress/wordpress-develop` rather than source it from `@wordpress/scripts` and maybe even allow to override in your project? This is the pattern we have implemented for all config files used in other scripts in `@wordpress/scripts` packages.
2. Why not include the latest WP instance in the docker and allow it override it somehow
this way it might work for core, plugins, and themes? E.g. in core, you would provide `--wp-dir=./` and that's all you would need. This way, you could easily opt-in for using `WordPress/wordpress-develop` if you want to use one env to rule them all but we would also support all those developers who are happy to use the latest WP distribution which doesn't require all the efforts to build it.

pento committed:
Reset the database before running e2e tests.

pento commented:
Thanks for the review, @gziolo! I know this PR has gotten fairly huge, so it's a bit of a tricky one to really get into. I think the easiest way to manage it is to get it to a point where we're comfortable that it works for most folks, then rapidly follow up with extra PRs as we get feedback. Given that this is the dev environment, feedback should be fairly rapid.

> It looks like this PR still needs some performance optimizations on Travis

Good point, I have some ideas to improve performance here, I'll try them out in a little bit.

> It looks like this could be simplified for Gutenberg developers and bundled in one command similar to what we have as of today with `./bin/setup-local-env.sh`.

I've updated the testing instructions to clarify that there are two methods to use it: `npm run env install` is the equivalent of `./bin/setup-local-env.sh`. The alternate method is for connecting to an existing `wordpress-develop` checkout, and is generally for more advanced debugging work.

> Why do you need to use `docker-compose.yml` from `WordPress/wordpress-develop` rather than source it from `@wordpress/scripts` and maybe even allow to override in your project?

There are a few reasons for this. It's worth noting that one of my primary guides for making decisions around `local-env` is to lay the groundwork for a dev environment that works as easily as possible for non-devs. With that in mind:

- Having a single `docker-compose.yml` file in Core, as opposed to having one in Core, one in Gutenberg, and many more in other feature plugins that will be able to use this, means that there's only set of Docker containers running. Docker can get fairly weird if you're trying to run containers with the same names, which mount volumes with the same names.
- Having multiple `docker-compose.yml` files doesn't lend itself to testing plugins together: each plugin ends up with its own environment which can't talk to others.
- Managing how plugins can mount older versions of WordPress becomes exponentially more difficult if the `docker-compose.yml` config needs to work for every WordPress branch (it needs to work with every branch, so that auto updates can be worked on), rather than being able to tweak the config for each branch.
- Ideally, feature plugins shouldn't need to do much customisation of the Docker environment, beyond mounting their particular volumes. Exposing the entire `docker-compose.yml` file is overkill.

> Why not include the latest WP instance in the docker and allow it override it somehow this way it might work for core, plugins, and themes?

This is a potential future iteration, but the infrastructure currently isn't in place for rebuilding Docker images with each Core commit.

pento committed:
Try using Travis build stages to cache the slow parts.

pento committed:
Tweak the Travis caching.

pento committed:
Try to share the cache between builds.

pento committed:
Speed up the cache warmup

pento committed:
Don't clone the repo if it's already cached.

pento committed:
Set the PHP version for PHP 5.6 tests.

pento committed:
Fix a .travis.yml parse error.

pento committed:
Give MySQL some time to start.

pento committed:
No need to cache wordpress/node_modules.

pento committed:
More Travis tweaking.

pento committed:
Try not caching node_modules

pento committed:
Try manually downloading the built version of WordPress.

pento committed:
Fix up some Travis bugs.

pento committed:
Add some debugging.

pento committed:
Travis: HALP HOW DO I mv

pento committed:
A little more tweaking.

pento committed:
Copy the tests directory from WP, too.

pento committed:
Improve performance of lint and license tests.

pento committed:
Add documentation to README.md.

pento commented:
I'm not wild about the hackery I had to do in `.travis.yml` to make it fast, but it works, soooo... 🤷🏻‍♂️

This addresses all comments and reviews on the PR to date.

pento committed:
Travis: Cache the cacheable bits of Composer.

gziolo commented:
> ### Method A (download and configure WordPress automagically)
> For testing the managed WordPress install, just run `npm run env install`, and watch the magic happen. 🎉

I love how this method :)

pento committed:
Update the @wordpress/scripts changelog.

shivapoudel commented:
@pento Since this is merged into master. Is there any way to install using `npm install` command. Something which I couldn't find right at the moment :)

pento commented:
@shivapoudel: Not yet: it'll be available when the next version of the `@wordpress/scripts` package is released. @gziolo might know when that's likely to happen, I'm not sure what the current plan is for increasing the frequency of package releases. 🙂

gziolo commented:
> Not yet: it'll be available when the next version of the `@wordpress/scripts` package is released. @gziolo might know when that's likely to happen, I'm not sure what the current plan is for increasing the frequency of package releases. 🙂

Usually, we publish to npm every time we publish the next version of Gutenberg plugin. This change should be available in 3 weeks unless there is a demand to cherry-pick this PR and do pre-release of the next version.

shivapoudel commented:
@gziolo Good enough to hear that with the release of Gutenberg 6.5, this package will get updated in npm repo 👍

aldavigdis commented:
For anyone, especially Linux users who are arriving here and have managed to run the original Gutenberg Docker setup with no issue until now and have the most recent version of the Docker engine installed via the official repository, you may need to update your version of `docker-compose` separately as well as the official Docker repo does not include it.

Go to https://github.com/docker/compose/releases and find the most recent non-RC release and follow the instructions.

I'm sorry for those who are package manager purists and feel like their setup is "dirty" when installing software like this, but it seems like the best way to go.

I'll follow-up on this by updating the docs as soon as I can.

pento commented:
Thank you for noting that @aldavigdis!

For reference, `docker-compose` version 1.22 is the minimum required: Fedora 29 provides this, but every other distro I looked at had older versions.

Unfortunately, it doesn't look like Docker are planning on providing DEB or RPM packages any time soon (https://github.com/docker/compose/issues/2235)

ellatrix commented:
I can no longer run e2e tests locally after this merge. `npm run env install -- --fast` is run when setting up e2e tests and gives the following error:

```
Warning: The 'gutenberg' plugin could not be found.
Error: No plugins activated.
```

The plugin is also gone when I load `http://localhost:8889/wp-admin/plugins.php`.

shivapoudel commented:
@pento Out of the context but don't get me wrong. Previously I use PHPMyAdmin image in docker-compose to access the MySQL database. Since I am new to the dockerized system if you do have way out how to access that container database using MySQL workbench or any other way which you prefer in the Windows environment during the development process will be a great help :)

pento commented:
@shivapoudel: You can run a PHPMyAdmin Docker container like so:

`docker run -d --link wordpress_mysql_1:db --net wordpress_wpdevnet -p 8081:80 phpmyadmin/phpmyadmin`

You may need to change `wordpress_mysql_1` to match the name of the MySQL Docker container, check `docker ps` for a list of running containers.

To login to PHPMyAdmin, username is `root`, password is `password`.

shivapoudel commented:
@pento That works like a charm. Thanks a lot!

shivapoudel commented:
@pento Since the env `LOCAL_DIR` is `src`, I get an error during new post/page creation. I have to change that to `build` and execute `npm run build` inside `wordpress` directory to resolve that issue. What is the easiest way for integration `env` so that I don't have to change that `LOCAL_DIR` to successfully add new post/page?

aduth commented:
There's many mixed references in code between usage as `npm run env:foo` and `npm run env foo`. Using the colon variant always results in an NPM error for me:

```
npm ERR! missing script: env:start
```

Should all of these be changed to be space-separated?

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/.travis.yml#L76

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/.travis.yml#L78

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/packages/scripts/scripts/env/cli.js#L17

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/packages/scripts/scripts/env/connect.js#L76

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/packages/scripts/scripts/env/connect.js#L77

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/packages/scripts/scripts/env/start.js#L30

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/packages/scripts/utils/env.js#L114

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/packages/scripts/utils/env.js#L123

https://github.com/WordPress/gutenberg/blob/f68a107c777169cd80b3ce56ec4f7c8360f2178c/packages/scripts/scripts/env/stop.js#L8

aduth commented on .travis.yml:
```diff
@@ -16,16 +16,79 @@ cache:
+
+      # Install WordPress.
+      cd wordpress
+      npm install dotenv wait-on
```
Not sure I understand the purpose or need for these installs.

`wait-on` seems like something which could have been useful in place of the `sleep 10` which follows a few lines down. I suppose it might have been an initial attempt to try to react to the availability of the environment in a more reliable way than a `sleep`?

aduth replied:
#18924

aduth commented on docs/contributors/getting-started.md:
```diff
@@ -2,74 +2,53 @@
-The easiest way to get started (on MacOS, Linux, or Windows 10 with the Linux Subsystem) is by running the Local Environment setup script, `./bin/setup-local-env.sh`. This will check if you have everything installed and updated, and help you download any extra tools you need.
+The first step is to install a recent version of Node. The easiest way (on MacOS, Linux, or Windows 10 with the Linux Subsystem) is by installing and running [nvm]. Once `nvm` is installed, you can install the correct version of Node by running `nvm install` in the Gutenberg directory.
 
-For another version of Windows, or if you prefer to set things up manually, be sure to have <a href="https://nodejs.org/en/">Node.js installed first</a>. You should be running a Node version matching the [current active LTS release](https://github.com/nodejs/Release#release-schedule) or newer for this plugin to work correctly. You can check your Node.js version by typing `node -v` in the Terminal ...
```
This was the only place we documented our exclusive support commitment to the active LTS. Was it intentional to change this? What versions of Node do we support now?

aduth replied:
#18923

aduth commented on packages/scripts/scripts/env/start.js:
```diff
@@ -0,0 +1,30 @@
+	// We're running in our managed WordPress environment, so check if we need to update.
+	let stat;
+	if ( false && commandExistsSync( 'git' ) ) {
+		stat = statSync( normalize( env.WP_DEVELOP_DIR + '/.git' ) );
```
This is dead code.

shivapoudel commented:
@pento I am receiving this kind of error during fast reinstall currently. Can you please do confirm :)

![image](https://user-images.githubusercontent.com/3774827/72842560-b833a080-3cc0-11ea-8306-647176c3be17.png)

mcsf commented:
Make sure you have Docker installed. The Gutenberg/WP environment expects `docker-compose` to be available.
